File: backend/src/modules/product/product.service.ts
import { Injectable, NotFoundException } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository, Like, Between } from 'typeorm';
import { Product } from './product.entity';
import { QueryProductDto } from './dto/query-product.dto';

@Injectable()
export class ProductService {
  constructor(
    @InjectRepository(Product)
    private productRepository: Repository<Product>,
  ) {}

  async findAll(queryDto: QueryProductDto) {
    const { page = 1, limit = 10, category_id, keyword, minPrice, maxPrice, sort } = queryDto;

    const queryBuilder = this.productRepository.createQueryBuilder('product');

    queryBuilder.where('product.status = :status', { status: 1 });

    if (category_id) {
      queryBuilder.andWhere('product.category_id = :category_id', { category_id });
    }

    if (keyword) {
      queryBuilder.andWhere('product.name LIKE :keyword', { keyword: `%${keyword}%` });
    }

    if (minPrice !== undefined) {
      queryBuilder.andWhere('product.price >= :minPrice', { minPrice });
    }

    if (maxPrice !== undefined) {
      queryBuilder.andWhere('product.price <= :maxPrice', { maxPrice });
    }

    switch (sort) {
      case 'price_asc':
        queryBuilder.orderBy('product.price', 'ASC');
        break;
      case 'price_desc':
        queryBuilder.orderBy('product.price', 'DESC');
        break;
      case 'sales':
        queryBuilder.orderBy('product.sales', 'DESC');
        break;
      case 'newest':
        queryBuilder.orderBy('product.created_at', 'DESC');
        break;
      default:
        queryBuilder.orderBy('product.sales', 'DESC');
    }

    queryBuilder.skip((page - 1) * limit).take(limit);

    const [items, total] = await queryBuilder.getManyAndCount();

    return {
      items,
      total,
      page,
      limit,
      totalPages: Math.ceil(total / limit),
    };
  }

  async findById(id: number): Promise<Product> {
    const product = await this.productRepository.findOne({
      where: { id, status: 1 },
      relations: ['category'],
    });

    if (!product) {
      throw new NotFoundException('商品不存在');
    }

    return product;
  }

  async findHot(limit: number = 8): Promise<Product[]> {
    return this.productRepository.find({
      where: { status: 1 },
      order: { sales: 'DESC' },
      take: limit,
    });
  }

  async findNew(limit: number = 8): Promise<Product[]> {
    return this.productRepository.find({
      where: { status: 1 },
      order: { created_at: 'DESC' },
      take: limit,
    });
  }
}

File: backend-spring/src/main/java/com/cakemall/module/product/ProductService.java
package com.cakemall.module.product;

import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.data.jpa.domain.Specification;
import org.springframework.stereotype.Service;

@Service
public class ProductService {

    private final ProductRepository productRepository;

    public ProductService(ProductRepository productRepository) {
        this.productRepository = productRepository;
    }

    public Page<Product> findAll(int page, int limit, Integer categoryId, String keyword, String sort) {
        Sort sorting;
        if ("price_asc".equals(sort)) {
            sorting = Sort.by(Sort.Direction.ASC, "price");
        } else if ("price_desc".equals(sort)) {
            sorting = Sort.by(Sort.Direction.DESC, "price");
        } else if ("sales".equals(sort)) {
            sorting = Sort.by(Sort.Direction.DESC, "sales");
        } else if ("newest".equals(sort)) {
            sorting = Sort.by(Sort.Direction.DESC, "createdAt");
        } else {
            sorting = Sort.by(Sort.Direction.DESC, "id");
        }

        Pageable pageable = PageRequest.of(page - 1, limit, sorting);

        Specification<Product> spec = Specification.where((root, query, cb) -> cb.equal(root.get("status"), 1));

        if (categoryId != null) {
            spec = spec.and((root, query, cb) -> cb.equal(root.get("categoryId"), categoryId));
        }

        if (keyword != null && !keyword.trim().isEmpty()) {
            spec = spec.and((root, query, cb) -> cb.like(root.get("name"), "%" + keyword.trim() + "%"));
        }

        return productRepository.findAll(spec, pageable);
    }

    public Product findById(Long id) {
        return productRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("商品不存在"));
    }

    public java.util.List<Product> findHot() {
        Pageable pageable = PageRequest.of(0, 8, Sort.by(Sort.Direction.DESC, "sales"));
        return productRepository.findByStatus(1, pageable).getContent();
    }

    public java.util.List<Product> findNew() {
        Pageable pageable = PageRequest.of(0, 8, Sort.by(Sort.Direction.DESC, "createdAt"));
        return productRepository.findByStatus(1, pageable).getContent();
    }
}
